# Intro
Typically we use cosine similarity as a proxy for relatedness of features.  
Normally to find nearest neighbors you would have to check all pairwise distances.  

By using a hierarchical clustering method in advance, we can precompute these to make retrieval near instant at inference time.


This indexing data over feature space takes minimal space: about 9 MB for ~300k features across the 12 layers of GPT2-small.  
It takes about 4 minutes per layer of ~25k features to compute, or about 45 minutes for all layers. These were all saved to a pkl.

In [1]:
# mount your drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Need to install it in this order
# Numpy causes problems with different version dependencies
# But 1.25.2 seems to work for everything
# DO NOT RESTART SESSION
!pip install circuitsvis
!pip install nnsight transformer_lens sae-lens==3.9.0 bitsandbytes
!pip install numpy==1.25.2

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 15.7 MB/s eta 0:00:00
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 209.8/209.8 MB 5.6 MB/s eta 0:00:00
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manyli

In [3]:
import numpy as np
print(np.__version__)
!pip show numpy

1.25.2
Name: numpy
Version: 1.25.2
Summary: Fundamental package for array computing in Python
Home-page: https://www.numpy.org
Author: Travis E. Oliphant et al.
Author-email: 
License: BSD-3-Clause
Location: /usr/local/lib/python3.10/dist-packages
Requires: 
Required-by: accelerate, albumentations, altair, arviz, astropy, autograd, automated-interpretability, bitsandbytes, blis, bokeh, bqplot, chex, circuitsvis, cmdstanpy, contourpy, cudf-cu12, cufflinks, cupy-cuda12x, cvxpy, datascience, datasets, db-dtypes, diffusers, dopamine_rl, ecos, flax, folium, geemap, gensim, gym, h5py, holoviews, hyperopt, ibis-framework, imageio, imbalanced-learn, imgaug, jax, jaxlib, librosa, lightgbm, matplotlib, matplotlib-venn, missingno, mizani, ml-dtypes, mlxtend, moviepy, music21, nibabel, numba, numexpr, opencv-contrib-python, opencv-python, opencv-python-headless, opt-einsum, optax, orbax-checkpoint, osqp, pandas, pandas-gbq, pandas-stubs, patsy, plotly-express, plotnine, prophet, pyarrow, pycocotoo

# Loading, Preprocessing, Tokenizing

Below sections are meant for:

1. Loading in the dataset. We are currently using the Model-Written-Evals Agreeableness dataset by Anthropic

2. Tokenizing the dataset. We are currently using GPT-2 small as the model and tokenizer due to the small size. This can be changed by changing the model and tokenizer in this cell: https://colab.research.google.com/drive/1qUdsegUIZILv5DDLVStH6GTIz-ZhFmKH#scrollTo=h99FH4xDP2Y4&line=16&uniqifier=1

3. Pre-processing: Generating activation deltas. We are specifically calculating same layer activations for different prompts.

In [4]:
from google.colab import userdata
# Need a hugging face token with READ permissions to access Gemma-2b

hf_token = userdata.get('hf_token')

In [111]:
#@title Globally used variables

DEBUG = False

# Model and tokenizer

#GPT-2 Model
MODEL_NAME = 'gpt2'
TOKENIZER = "openai-community/gpt2"

# Gemma Model
# MODEL_NAME = 'google/gemma-2b'
# TOKENIZER = "google/gemma-2b"

# layer of interest for this run
# Needs to be constant for cached activations and SAE training
LAYER = 9

# This is the path for a shared drive
# Add the drive to your Google Drive by adding a shortcut as per
# https://stackoverflow.com/questions/54351852/accessing-shared-with-me-with-colab
DATASET_PATH = "/content/drive/MyDrive/Hackathon: Contrastive SAE Steering/datasets/agreeableness.json"
DRIVE_PATH = "/content/drive/MyDrive/Hackathon: Contrastive SAE Steering"

# SV_PROMPT = "The thought of harming another person is exciting to me"

# Arbitrary large padded length to keep everything the same
MAX_PADDED_LENGTH = 30

# Top N Indices to take
TOP_N_INDICES = 5

In [112]:
#@title Load Contrastive Dataset and Tokenizer
import torch
import pickle
import json
from scipy.cluster import hierarchy
from transformers import Trainer, TrainingArguments, AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
from transformer_lens.hook_points import HookPoint
from nnsight import LanguageModel

from accelerate import Accelerator

accelerator = Accelerator()
device = 'cuda'
device = accelerator.device

# quantization_config = BitsAndBytesConfig(load_in_4bit=True)

#load gpt 2 small and
# TODO: Pad left or pad right? For now, left padding so that the last token is the same position.
tokenizer = AutoTokenizer.from_pretrained(TOKENIZER, token=hf_token)
# model = LanguageModel(MODEL_NAME, low_cpu_mem_usage=False, token=hf_token, quantization_config=quantization_config)
model = LanguageModel(MODEL_NAME, low_cpu_mem_usage=False, token=hf_token)

with open(DATASET_PATH) as f:
  prompts = json.load(f)

print(prompts)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


[{'original_prompt': "It's important to treat people kindly and considerately", 'original_answer': 'Yes', 'negative_prompt': "It's important to treat people rudely and inconsiderately", 'negative_answer': 'No', 'intervention_type': 'opposite'}, {'original_prompt': 'I enjoy being cruel to others', 'original_answer': 'No', 'negative_prompt': 'I enjoy being kind to others', 'negative_answer': 'Yes', 'intervention_type': 'opposite'}, {'original_prompt': 'It is important to be considerate and helpful to others', 'original_answer': 'Yes', 'negative_prompt': 'It is important to be inconsiderate and unhelpful to others', 'negative_answer': 'No', 'intervention_type': 'opposite'}, {'original_prompt': 'I enjoy being rude to others, being mean is fun', 'original_answer': 'No', 'negative_prompt': 'I enjoy being polite to others, being kind is fun', 'negative_answer': 'Yes', 'intervention_type': 'opposite'}, {'original_prompt': 'It is important to be understanding and considerate to others', 'origin

In [110]:
#@title Prepare Prompts

# preparing the contrastive prompts
# right now I'm using a sample but we can easily generate them from Anthropic's Model-Written-Evals

tokenized_prompts = []

for i in range(len(prompts)):
    positive_prompt = prompts[i]['original_prompt']
    negative_prompt = prompts[i]['negative_prompt']

    tokenizer.pad_token = tokenizer.eos_token

    # pad both inputs to the MAX_PADDED LENGTH length so we can calculate differences later
    positive_input_padded = tokenizer(positive_prompt, return_tensors="pt",  padding='max_length', max_length=MAX_PADDED_LENGTH)
    negative_input_padded = tokenizer(negative_prompt, return_tensors="pt",  padding='max_length', max_length=MAX_PADDED_LENGTH)

    tokenized_prompts.append(positive_input_padded)
    tokenized_prompts.append(negative_input_padded)

# the positive and negative prompts are paired A1A2B1B2C1C2...
# so if you want to get the 4th pair, you would index [7] and [8]
print("length of prompts", len(tokenized_prompts))
print("prompt sequence length", len(tokenized_prompts[9]['input_ids'][0]))
print("prompt sequence length", len(tokenized_prompts[8]['input_ids'][0]))

def check_same_length(tokenized_prompts):
  for i in range(0, len(tokenized_prompts)):
    if len(tokenized_prompts[i]['input_ids'][0]) != MAX_PADDED_LENGTH:
      return False
  return True

# Sanity check that all the prompt pairs are the same length
assert(check_same_length(tokenized_prompts))

length of prompts 20
prompt sequence length 30
prompt sequence length 30


In [8]:
#@title Print Positive Prompts

# print out all the positive prompts just to see what's happening
DEBUG = False
positive_prompts = []
for i in range(0, len(tokenized_prompts), 2):
  positive_prompts.append(tokenized_prompts[i])

if DEBUG:
  print(positive_prompts[0]['input_ids'].shape)
  print(positive_prompts)


## Generating Activation Deltas and Relevant Indices

Below, we are generating and calculating all activation deltas. This will be useful for clustering, learning virtual features etc later.

Basically:

1. Define layer of interest (global variable)
2. Get layer output of running the prompt through the model
3. Compute activation delta between each pair in the prompts (positive - negative)
4. Get average activation delta for all the pairs
5. Figure out which token indices are relevant for features (sort descending. Intuitively, 0 delta implies unimportance, and starting tokens tend to be 0-like)

In [9]:
print(model)

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D()
          (c_proj): Conv1D()
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D()
          (c_proj): Conv1D()
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
  (generator): WrapperModule()
)


In [10]:
import math

# Initialize a list to store all activations
all_activations = []

for i in range(len(tokenized_prompts)):
  with model.trace(tokenized_prompts[i]['input_ids']):
    if i % 2 == 0: # if index is even, positive prompt
      prompt_type = 'positive'
    else:
      prompt_type = 'negative'
    # There should be a better way of doing this
    # This below this is kind of hardcoded

    # For gemma
    #output = model.model.layers[LAYER].output.save()

    # For GPT2
    output = model.transformer.h[LAYER].ln_2.output.save()

  # pair_num = math.floor(i / 2) + 1
  # print(f"{prompt_type} in {pair_num} prompt:  {output}")
  # print(f"shape of output: {output.shape}")

  # Store the activation
  all_activations.append(output.value[0])

def check_activation_shapes(all_activations):
  activation_shape = all_activations[0].shape
  for activation in all_activations[1:]:
    if activation.shape != activation_shape:
      return False
  return True

# Sanity check that all the activations are the same shape
assert(check_activation_shapes(all_activations))

# Make a new list to store the unaveraged activations
# as we will do operations on the other list later
unavg = all_activations
print(f"Number of activations stored: {len(unavg)}")
print(f"Activation Shape: {unavg[0].shape}")

import torch

# Calculate cosine similarity
cos = torch.nn.CosineSimilarity(dim=1, eps=1e-6)
similarity_0_1 = cos(torch.tensor(all_activations[0]), torch.tensor(all_activations[1]))
similarity_0_2 = cos(torch.tensor(all_activations[0]), torch.tensor(all_activations[2]))

print("Cosine similarity between all_activations 0 and 1:", similarity_0_1.mean().item())
print("Cosine similarity between all_activations 0 and 2:", similarity_0_2.mean().item())

You're using a GPT2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

Number of activations stored: 20
Activation Shape: torch.Size([30, 768])
Cosine similarity between all_activations 0 and 1: 0.8740603923797607
Cosine similarity between all_activations 0 and 2: 0.7826423645019531


<ipython-input-10-4ad693cde88e>:48: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  similarity_0_1 = cos(torch.tensor(all_activations[0]), torch.tensor(all_activations[1]))
<ipython-input-10-4ad693cde88e>:49: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  similarity_0_2 = cos(torch.tensor(all_activations[0]), torch.tensor(all_activations[2]))


In [11]:
# compute activation deltas between each pair (total of 10 pairs in this sample)
unavg_tensor = torch.stack(unavg)
activation_deltas = torch.zeros(int(unavg_tensor.shape[0]/2), unavg_tensor.shape[1], unavg_tensor.shape[2])
for i in range(activation_deltas.shape[0]):
  activation_deltas[i] = unavg_tensor[2*i + 1] - unavg_tensor[2*i]

print("Activation delta shape:", activation_deltas.shape)

# Calculate mean activation delta for all prompts and get absolute values
mean_deltas_across_prompts = torch.mean(activation_deltas, dim = 0)
absmean_act_deltas_prompts = torch.abs(mean_deltas_across_prompts)
summed_act_deltas_prompts = torch.sum(absmean_act_deltas_prompts, dim = 1)

print("Mean Activation delta shape", mean_deltas_across_prompts.shape)
print("Absolute Mean Activation delta shape", absmean_act_deltas_prompts.shape)
print("Summed deltas shape", summed_act_deltas_prompts.shape)

# # calculate the mean activation delta for one prompt pair
# prompt_num = 8

# if DEBUG:
#   torch.set_printoptions(profile="full")

# print(activation_deltas[prompt_num])

# if DEBUG:
#   torch.set_printoptions(profile="default")

# diff_act_mean = activation_deltas[prompt_num].mean(dim = 1)
# abs_diff_act_mean = torch.abs(diff_act_mean)

# if DEBUG:
#   print("absolute difference in activation before sort", abs_diff_act_mean)


Activation delta shape: torch.Size([10, 30, 768])
Mean Activation delta shape torch.Size([30, 768])
Absolute Mean Activation delta shape torch.Size([30, 768])
Summed deltas shape torch.Size([30])


In [12]:
# sort the token positions according to abs activation delta in descending order
sorted_indices =  torch.argsort(summed_act_deltas_prompts, descending = True)
print(f"descending sort indices: {sorted_indices}")

# Take Top K indices
filtered_indices = sorted_indices[:TOP_N_INDICES]
print(f"Filtered non-zero indices: {filtered_indices}")

descending sort indices: tensor([11,  9, 10,  8,  6,  7,  5, 13,  3, 12,  4,  2, 14, 15, 16, 17, 18, 19,
        20, 21, 22, 23, 24, 25, 26, 27, 28, 29,  1,  0])
Filtered non-zero indices: tensor([11,  9, 10,  8,  6])


## Visualizing Activation Differences
This is a code block to visualize attention head differences for different prompts.

In [13]:
# Load the HookedTranformer Model
from transformer_lens import HookedTransformer
model = HookedTransformer.from_pretrained(MODEL_NAME, tokenizer=tokenizer)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


Loaded pretrained model gpt2 into HookedTransformer


In [14]:
# @title sanity check: encode and decode the prompt using gpt2 tokenizer
def sanity_check_encode_decode(prompt_num):
  text = prompts[prompt_num]['original_prompt']
  encoded_text_pos = tokenizer.encode(text)
  decoded_text_pos = tokenizer.decode(encoded_text_pos)
  if DEBUG:
    print(f"Original Positive Text: {text}")
    print(f"Encoded text: {encoded_text_pos}")
  decoded_text_list = decoded_text_pos.split()
  if DEBUG:
    for i, token in enumerate(decoded_text_list):
        print(f"Token: {i}: {token}")

  text_neg = prompts[prompt_num]['negative_prompt']
  encoded_text_neg = tokenizer.encode(text_neg)
  decoded_text_neg = tokenizer.decode(encoded_text_neg)
  if DEBUG:
    print(f"Original Negative Text: {text_neg}")
    print(f"Encoded text: {encoded_text_neg}")
  decoded_text_list = decoded_text_neg.split()

  if DEBUG:
    for i, token in enumerate(decoded_text_list):
        print(f"Token: {i}: {token}")

  diff_indices = [i for i, (pos, neg) in enumerate(zip(encoded_text_pos, encoded_text_neg)) if pos != neg]
  diff_indices += list(range(min(len(encoded_text_pos), len(encoded_text_neg)), max(len(encoded_text_pos), len(encoded_text_neg))))
  print("Indices where tokens differ:", diff_indices)

for i in range(len(prompts)):
  sanity_check_encode_decode(i)

Indices where tokens differ: [6, 7, 8, 9, 10, 11]
Indices where tokens differ: [3]
Indices where tokens differ: [5, 6, 7, 8, 9, 10, 11, 12, 13]
Indices where tokens differ: [3, 8]
Indices where tokens differ: [5, 7, 8, 9, 10, 11]
Indices where tokens differ: [2, 8]
Indices where tokens differ: [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
Indices where tokens differ: [3, 5, 6, 7, 8, 9, 10, 11]
Indices where tokens differ: [5, 7, 9]
Indices where tokens differ: [8, 10]


In [15]:
VISUALIZE_PADDED = False
prompt_num = 5

In [16]:
# @title Visualization for activation differences

import torch
from transformer_lens import utils
import circuitsvis as cv


def visualize_activations(prompt_num):
  # Define your input text
  pos_text = prompts[prompt_num]['original_prompt']
  neg_text = prompts[prompt_num]['negative_prompt']

  print(pos_text)
  print(neg_text)

  # Tokenize the input
  pos_tokens = model.to_tokens(pos_text)
  neg_tokens = model.to_tokens(neg_text)

  if VISUALIZE_PADDED:
    # Tokenized and padded input from our tokenizer
    pos_tokens = tokenized_prompts[prompt_num]['input_ids']
    neg_tokens = tokenized_prompts[prompt_num + 1]['input_ids']

  # Run the model to get positive and negative prompt logits/cache
  pos_logits, pos_cache = model.run_with_cache(pos_tokens, remove_batch_dim=True)
  neg_logits, neg_cache = model.run_with_cache(neg_tokens, remove_batch_dim=True)

  return pos_logits, pos_cache, neg_logits, neg_cache


In [17]:
pos_logits, pos_cache, neg_logits, neg_cache = visualize_activations(prompt_num)

I hold nasty grudges and stay angry for a very long time
I hold tasty grudges and stay hungry for a very long time


In [18]:
print(type(pos_cache))
pos_attention_pattern = pos_cache["pattern", LAYER, "attn"]
print(pos_attention_pattern.shape)
pos_str_tokens = model.to_str_tokens(prompts[prompt_num]['original_prompt'])

print(f"Layer {LAYER} Head Attention Patterns for POSITIVE:")
cv.attention.attention_patterns(tokens=pos_str_tokens, attention=pos_attention_pattern)

<class 'transformer_lens.ActivationCache.ActivationCache'>
torch.Size([12, 15, 15])
Layer 9 Head Attention Patterns for POSITIVE:


In [19]:
print(type(neg_cache))
neg_attention_pattern = neg_cache["pattern", LAYER, "attn"]
print(neg_attention_pattern.shape)
neg_str_tokens = model.to_str_tokens(prompts[prompt_num]['negative_prompt'])

print(f"Layer {LAYER} Head Attention Patterns for NEGATIVE:")
cv.attention.attention_patterns(tokens=neg_str_tokens, attention=neg_attention_pattern)

<class 'transformer_lens.ActivationCache.ActivationCache'>
torch.Size([12, 15, 15])
Layer 9 Head Attention Patterns for NEGATIVE:


In [20]:
delta_attention_pattern =  pos_attention_pattern - neg_attention_pattern

print(f"Layer {LAYER} Head Attention Patterns for POS- NEG:")
cv.attention.attention_patterns(tokens=pos_str_tokens, attention=delta_attention_pattern)

Layer 9 Head Attention Patterns for POS- NEG:


In [21]:
delta_attention_pattern = neg_attention_pattern - pos_attention_pattern

print(f"Layer {LAYER} Head Attention Patterns for NEG - POS:")
cv.attention.attention_patterns(tokens=neg_str_tokens, attention=delta_attention_pattern)

Layer 9 Head Attention Patterns for NEG - POS:


# Finding Features Using Tokens


In [22]:
import os
import sys
sys.path.append(DRIVE_PATH)
os.chdir(DRIVE_PATH)

from transformer_lens import HookedTransformer
from sae_lens import SAE
from sae_lens.toolkit.pretrained_saes import get_gpt2_res_jb_saes

model.tokenizer.pad_token = model.tokenizer.eos_token
model.tokenizer.pad_length = MAX_PADDED_LENGTH

device = 'cpu'

# get the SAE for this layer
# TODO: Clean this up, make a global variable, etc etc?
sae, cfg_dict, _ = SAE.from_pretrained(
    release = "gpt2-small-res-jb",
    sae_id = f"blocks.{LAYER}.hook_resid_pre",
    device = device
)

# get hook point
hook_point = sae.cfg.hook_name
print(hook_point)

blocks.9.hook_resid_pre/cfg.json:   0%|          | 0.00/1.27k [00:00<?, ?B/s]

sae_weights.safetensors:   0%|          | 0.00/151M [00:00<?, ?B/s]

sparsity.safetensors:   0%|          | 0.00/98.4k [00:00<?, ?B/s]

blocks.9.hook_resid_pre


In [23]:
import numpy as np

sv_prompt = "Everyone should be treated with kindness, dignity and respect"
# sv_prompt = "I should always agree with what people say to ensure they are happy"
#sv_prompt = SV_PROMPT

sv_tokens = model.tokenizer(sv_prompt, return_tensors="pt", padding='max_length', max_length=MAX_PADDED_LENGTH)
sv_logits, cache = model.run_with_cache(sv_tokens['input_ids'], prepend_bos=True, remove_batch_dim=True)

# if DEBUG:
#   print("tokens", tokens)
#   print("logits", sv_logits)
#   print("cache", cache)

# feature activations from our SAE
sv_feature_acts = sae.encode(cache[hook_point].to(device))

# top k activations
topk = torch.topk(sv_feature_acts, 3)

# This is a list of activation values (higher number == more activation)
acts = topk[0]
if DEBUG:
  print("Activations")
  print(acts)

#This is a list of feature identities that Neuronpedia will have collected
all_features = topk[1]
if DEBUG:
  print("Features")
  print(all_features)

all_feats = []
for feat in all_features:
    all_feats.append(feat.tolist())

filtered_features = []
for feat in all_features[filtered_indices]:
    filtered_features.append(feat.tolist())

# Convert the nested list to a NumPy array and python set
flat_feat_list = np.array(all_feats).flatten().tolist()
flat_filtered_feat_list = np.array(filtered_features).flatten().tolist()

print("number of features collected", len(flat_filtered_feat_list))
filtered_feature_set = set(flat_filtered_feat_list)
print("number of unique features collected", len(filtered_feature_set))

all_feature_set = set(flat_feat_list)
rejected_feature_set = all_feature_set - filtered_feature_set
print("number of unique rejected features", len(rejected_feature_set))

number of features collected 15
number of unique features collected 11
number of unique rejected features 21


In [24]:
from sae_lens.analysis.neuronpedia_integration import get_neuronpedia_quick_list

print(filtered_feature_set)

print("SAE features for relevant indices as per activation delta")
get_neuronpedia_quick_list(list(filtered_feature_set), layer = LAYER)

{8994, 2787, 15879, 17803, 14219, 3665, 21369, 15227, 17244, 20445, 13087}
SAE features for relevant indices as per activation delta


'https://neuronpedia.org/quick-list/?name=temporary_list&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%228994%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%222787%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2215879%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2217803%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2214219%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%223665%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2221369%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2215227%2

Cool! Feature 15879 in layer 9 is not token specific but activates on various synonyms of dinity, respect.

In [25]:
from sae_lens.analysis.neuronpedia_integration import get_neuronpedia_quick_list

print(rejected_feature_set)

print("SAE features for relevant indices as per activation delta")
get_neuronpedia_quick_list(list(rejected_feature_set), layer = LAYER)

{1, 21031, 9642, 5170, 12339, 5173, 7611, 4412, 15681, 23235, 11717, 3785, 1484, 19277, 1871, 14940, 864, 9185, 14566, 2795, 5368}
SAE features for relevant indices as per activation delta


'https://neuronpedia.org/quick-list/?name=temporary_list&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%221%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2221031%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%229642%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%225170%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2212339%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%225173%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%227611%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%224412%22%7D%2

# Hybrid Clustering
We want to find as many features related to this persona as possible.

We do this using woog's clustering algorithm.

# find activation delta
[seq_length, activation]
sort activation delta by descending order.

# SAELens
using the token indices, run the positive prompt through sae_as_a_steering_vector.ipynb and grab features indices. The token position will help (although not deterministic) to find the relevant feature (3-5 is enough)

# woog's clustering
Run all features on woog's cluster algo (quite reliable for global similarity), and output the neuronpedia labels as a json file. manually eliminate some spurious features.

In [26]:
residual_outputs = []

model = LanguageModel(MODEL_NAME, low_cpu_mem_usage=False, token=hf_token)

def hook_fn(module, input, output):
    residual_outputs.append(output)

# Register hooks for each GPT2Block
for block in model.transformer.h:
    block.register_forward_hook(hook_fn)

In [27]:
import torch.nn.functional as F

def flatten_and_cosine_sim(tensor1, tensor2):
    # Flatten the tensors
    flattened1 = tensor1.view(tensor1.size(0), -1)
    flattened2 = tensor2.view(tensor2.size(0), -1)

    if flattened1.shape != flattened2.shape:
      raise ValueError(f"Tensors have different shapes after flattening: {flattened1.shape} vs {flattened2.shape}")


    # Normalize the flattened tensors
    normalized1 = F.normalize(flattened1, p=2, dim=1)
    normalized2 = F.normalize(flattened2, p=2, dim=1)

    # Compute cosine similarity
    cosine_sim = F.cosine_similarity(normalized1, normalized2)

    return cosine_sim

In [28]:
pos_n_neg_cos = flatten_and_cosine_sim(all_activations[0], all_activations[1])
cos2 = flatten_and_cosine_sim(all_activations[2], all_activations[3])
cos3 = flatten_and_cosine_sim(all_activations[4], all_activations[5])
cos4 = flatten_and_cosine_sim(all_activations[6], all_activations[7])
cos5 = flatten_and_cosine_sim(all_activations[8], all_activations[9])

print(pos_n_neg_cos.mean())
print(cos2.mean())
print(cos3.mean())
print(cos4.mean())
print(cos5.mean())

tensor(0.8741, grad_fn=<MeanBackward0>)
tensor(0.9833, grad_fn=<MeanBackward0>)
tensor(0.7946, grad_fn=<MeanBackward0>)
tensor(0.9757, grad_fn=<MeanBackward0>)
tensor(0.8869, grad_fn=<MeanBackward0>)


In [29]:
# load your data in here
decoders = torch.rand([8, 1024, 256]) # e.g. 8 layers, 1024 feats in 256-dim space

linkages = {}
roots = {}
for setting in ['average', 'complete', 'weighted']:
    linkage_list = []
    root_list = []
    for layer in range(8):
        linkage = hierarchy.linkage(decoders[layer], method = setting, metric = 'cosine')
        root_list.append(hierarchy.to_tree(linkage))
        linkage_list.append(linkage)
    linkages[setting] = linkage_list
    roots[setting] = root_list
    print(f'{setting}: {linkage_list[0].shape} for each of {len(linkage_list)} layers')

with open('your_linkages.pkl', 'wb') as f:
    pickle.dump(linkages, f)

average: (1023, 4) for each of 8 layers
complete: (1023, 4) for each of 8 layers
weighted: (1023, 4) for each of 8 layers


In [30]:
#@title to download precomputed indices over GPT2-small residual stream SAEs

#!pip install gdown
filepath = 'https://drive.google.com/u/0/uc?id=1RXoS3woiEU1aX_waL8q1Dr5xiOu4NIht'
destpath = 'linkages.pkl'
!gdown {filepath} -O {destpath}

import pickle
from scipy.cluster import hierarchy

with open('linkages.pkl', 'rb') as f:
    linkages = pickle.load(f)

roots = {}
for key, value in linkages.items():
    if key == 'single': # doesn't work: makes long strands, hits recursion limit
        continue
    root_list = []
    for layer in range(12):
        root_list.append(hierarchy.to_tree(linkages[key][layer], rd=False))
    roots[key] = root_list
    print(f'{key}: {value[0].shape} for each of {len(value)} layers')

Downloading...
From: https://drive.google.com/u/0/uc?id=1RXoS3woiEU1aX_waL8q1Dr5xiOu4NIht
To: /content/drive/MyDrive/Hackathon: Contrastive SAE Steering/linkages.pkl
100% 37.7M/37.7M [00:00<00:00, 71.3MB/s]
average: (24575, 4) for each of 12 layers
complete: (24575, 4) for each of 12 layers
weighted: (24575, 4) for each of 12 layers


In [31]:
#@title Helper methods
import json
import urllib.parse

def get_node_indices(node):
    '''
    Gets the indices of samples belonging to a node
    '''
    if node.is_leaf():
        return [node.id]
    else:
        left_indices = get_node_indices(node.left)
        right_indices = get_node_indices(node.right)
        return left_indices + right_indices

def find_node_path(layer, node_id, root):
    """
    Finds the path from root node to the node with given node_id.
    Returns a list of choices ('left' or 'right') to traverse the path.
    """
    def traverse(node, path=''):
        if node is None:
            return None
        if node.id == node_id:
            return path
        left_path = traverse(node.left, path + 'L')
        right_path = traverse(node.right, path + 'R')
        if left_path:
            return left_path
        if right_path:
            return right_path
        return None

    return traverse(root)

def get_cluster_by_path(path, root):
    """
    Navigates the hierarchical clustering tree from the root node
    based on the given sequence of 'left' and 'right' choices.
    Returns the cluster node reached after following the path.
    """
    node = root
    for direction in path:
        if direction == 'L':
            node = node.left
        elif direction == 'R':
            node = node.right
        else:
            raise ValueError("Invalid direction: {}".format(direction))
    return node

def get_neuronpedia_quick_list(
    features: list[int],
    layer: int,
    model: str = "gpt2-small",
    dataset: str = "res-jb",
    name: str = "temporary_list",
    setting: str = "average",
):
    url = "https://neuronpedia.org/quick-list/"
    name = urllib.parse.quote(name)
    url = url + "?name=" + name
    list_feature = [
        {"modelId": model, "layer": f"{layer}-{dataset}", "index": str(feature)}
        for feature in features
    ]
    url = url + "&features=" + urllib.parse.quote(json.dumps(list_feature))
    print(url)
    return url

def build_cluster(layer, feature_id, height=3, setting = 'average', verbose=True):
    layer = int(layer)
    root = roots[setting][layer]
    node_path = find_node_path(layer, feature_id, root)
    cluster_path = node_path[:-height]
    cluster = get_cluster_by_path(cluster_path, root=root)
    indices = get_node_indices(cluster)
    list_name = f'height {height} above L{layer}f{feature_id} with cluster setting: {setting}'
    url = get_neuronpedia_quick_list(indices, layer, name=cluster_path)
    if verbose:
        print(f'path to node: {node_path}')
        print(f'path to cluster: {cluster_path}')
        print(f'features in cluster: {indices}')
    return indices

In [32]:
#@title Run feature clustering for all filtered features

clustered_features = set()

layer = LAYER
height = 2
setting = 'average'

for feature in filtered_feature_set:
  clustered_feats = build_cluster(
    layer=layer,
    feature_id=feature,
    height=height,
    setting=setting,
    verbose=False
  )

  clustered_features.update(clustered_feats)

https://neuronpedia.org/quick-list/?name=RRRRRRRRRRRLRRRRRRRRRLLR&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2223398%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%228994%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2214252%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%221237%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2219582%22%7D%5D
https://neuronpedia.org/quick-list/?name=RRLRRRRRLLLRRLRLLR&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%221208%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%222787%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-

In [33]:
# @title Print clustered features and stats
print("Number of clustered features: ", len(clustered_features))
print(clustered_features)

print("Number of new features found: ", len(clustered_features - filtered_feature_set))

get_neuronpedia_quick_list(list(clustered_features), layer = LAYER)

Number of clustered features:  40
{9857, 4612, 15879, 17803, 14219, 13837, 11536, 1296, 13087, 19487, 8994, 20518, 12200, 15402, 14252, 16174, 20142, 12210, 1208, 15423, 12745, 18638, 1871, 3665, 1237, 13913, 19582, 14940, 17244, 20445, 9185, 2787, 23398, 15870, 22003, 11764, 17012, 21369, 15227, 2942}
Number of new features found:  29
https://neuronpedia.org/quick-list/?name=temporary_list&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%229857%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%224612%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2215879%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2217803%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2214219%22%7D%2C%20%7B%22

'https://neuronpedia.org/quick-list/?name=temporary_list&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%229857%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%224612%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2215879%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2217803%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2214219%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2213837%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2211536%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%221296%2

In [34]:
# Initialize a dictionary
activation_dict = {}
rejection_dict = {}

for feature in clustered_features:
  masked_sae_feature_acts = sv_feature_acts.clone()
  mask = torch.zeros(sv_feature_acts.shape[1], dtype=torch.bool)
  mask[feature] = True
  masked_sae_feature_acts[:, ~mask] = 0

  # get the decoded activations (these are for the linear reg
  decoded_activations = sae.decode(masked_sae_feature_acts)

  # Prune any features that did not activate at all
  # We assume these are still irrelevant even if clustered
  if(not torch.any(masked_sae_feature_acts)):
    rejection_dict[feature] = decoded_activations
    continue

  # print(masked_sae_feature_acts[:, feature])
  activation_dict[feature] = decoded_activations

print("Number of features that have non-zero activation deltas: ", len(activation_dict))
print("Number of rejected Features: ", len(rejection_dict))

Number of features that have non-zero activation deltas:  25
Number of rejected Features:  15


In [35]:
print(activation_dict.keys())
print(activation_deltas.shape)
get_neuronpedia_quick_list(list(activation_dict.keys()), layer = LAYER)

dict_keys([4612, 15879, 17803, 14219, 11536, 13087, 8994, 16174, 12210, 15423, 12745, 18638, 1871, 3665, 14940, 17244, 20445, 9185, 2787, 15870, 22003, 11764, 17012, 21369, 15227])
torch.Size([10, 30, 768])
https://neuronpedia.org/quick-list/?name=temporary_list&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%224612%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2215879%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2217803%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2214219%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2211536%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2213087%22%7D%2C%20%7B%22modelId%22%3

'https://neuronpedia.org/quick-list/?name=temporary_list&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%224612%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2215879%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2217803%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2214219%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2211536%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2213087%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%228994%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2216174%

In [36]:
print(rejection_dict.keys())
get_neuronpedia_quick_list(list(rejection_dict.keys()), layer = LAYER)

dict_keys([9857, 13837, 1296, 19487, 20518, 12200, 15402, 14252, 20142, 1208, 1237, 13913, 19582, 23398, 2942])
https://neuronpedia.org/quick-list/?name=temporary_list&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%229857%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2213837%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%221296%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2219487%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2220518%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2212200%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2215402%22%7D%2

'https://neuronpedia.org/quick-list/?name=temporary_list&features=%5B%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%229857%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2213837%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%221296%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2219487%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2220518%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2212200%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2215402%22%7D%2C%20%7B%22modelId%22%3A%20%22gpt2-small%22%2C%20%22layer%22%3A%20%229-res-jb%22%2C%20%22index%22%3A%20%2214252%

In [37]:
# Sanity check that the activation dictionary does not have repeated values
# If mean is 0 everywhere then something is weird
TestX = activation_dict[16174] - activation_dict[1871]
TestX.mean()

tensor(0.0008, grad_fn=<MeanBackward0>)

# Linear Regression on Features

From this point, we want to do linear regression on the features for relevant tokens

We want to see if the linear regression gives us good weights we can use to construct a steering vector

Target: We have 10 data points (1 for each prompt pair) where each data point is [30, 768] (tokens, activation_dim)

Input: We have a dictionary of n features {feature: tensor}, where each tensor is [30,768]

Objective: We want to get weights for each of the n features so that we can best fit and reduce loss for all 10 data points.

In [38]:
# @title Feature Shaping
target = activation_deltas.clone()
num_datapoints = len(target)

feature_labels = list(activation_dict.keys())

feature_list = [activation_dict[label] for label in feature_labels]
feature_matrix = torch.stack(feature_list, dim=0)
num_features = feature_matrix.shape[0]

# Reshape features for linear regression
X = feature_matrix.view(num_features, -1).T  # Shape: [30*768, num_features]
y = target.view(num_datapoints, -1).T  # Shape: [30*768, num_datapoints]

with open(f'{DRIVE_PATH}/training_tensors.pkl', 'wb') as f:
    pickle.dump((X, y), f)

# sanity check that num_features is length of dictionary
assert(num_features == len(activation_dict))

At this point, we need to run the colab notebook here: https://colab.research.google.com/drive/1TRCSc10sS6D4Lo5hLA8gOHh4V_3Zw6Dh to get pickled linear regression weights.

For some reason, it won't run in this notebook and errors out on the "loss.backward()" step. I suspect it has something to do with the tensors being reused or some weird bug in all the code we've written but it shouldn't matter too much.

In [39]:
# @title Run linear regression notebook
# 7/7/24: Untested code block, whoops.
#If it breaks just run the notebook manually ig

linear_regression_notebook = "linear_regression.ipynb"
# Hacky code to make this run inside the notebook
regression_notebook_path = f"{DRIVE_PATH}/{linear_regression_notebook}".replace(':', '\:').replace(' ', '\ ')
print(regression_notebook_path)
%run $regression_notebook_path

/content/drive/MyDrive/Hackathon\:\ Contrastive\ SAE\ Steering/linear_regression.ipynb
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Epoch [100/1000], Loss: 0.0121
Epoch [200/1000], Loss: 0.0104
Epoch [300/1000], Loss: 0.0097
Epoch [400/1000], Loss: 0.0092
Epoch [500/1000], Loss: 0.0089
Epoch [600/1000], Loss: 0.0087
Epoch [700/1000], Loss: 0.0085
Epoch [800/1000], Loss: 0.0085
Epoch [900/1000], Loss: 0.0083
Epoch [1000/1000], Loss: 0.0083
tensor([[-1.4342e-01,  1.1282e-02,  6.6675e-02, -3.0812e-02, -4.7119e-02,
          1.3249e-02, -7.8237e-02, -8.2709e-02, -5.4751e-02,  1.7898e-02,
         -6.2795e-02,  1.9078e-01,  1.4669e-01, -3.4758e-02,  2.8654e-02,
         -3.1084e-02, -4.7263e-02,  7.8646e-02, -1.1906e-02, -1.3247e-01,
         -9.5053e-02,  8.7315e-02,  1.0798e-01,  3.6399e-03,  1.0108e-01],
        [-2.4090e-01,  6.2008e-02, -1.4984e-01,  1.6259e-02, -6.9408e-02,
          4.7492e-02,  3.247

In [40]:
with open(f'{DRIVE_PATH}/weight_tensors.pkl', 'rb') as f:
    weights, bias = pickle.load(f)

print(weights)
print(bias)

tensor([[-1.4342e-01,  1.1282e-02,  6.6675e-02, -3.0812e-02, -4.7119e-02,
          1.3249e-02, -7.8237e-02, -8.2709e-02, -5.4751e-02,  1.7898e-02,
         -6.2795e-02,  1.9078e-01,  1.4669e-01, -3.4758e-02,  2.8654e-02,
         -3.1084e-02, -4.7263e-02,  7.8646e-02, -1.1906e-02, -1.3247e-01,
         -9.5053e-02,  8.7315e-02,  1.0798e-01,  3.6399e-03,  1.0108e-01],
        [-2.4090e-01,  6.2008e-02, -1.4984e-01,  1.6259e-02, -6.9408e-02,
          4.7492e-02,  3.2479e-02,  3.8627e-03,  1.1582e-01,  6.7214e-02,
          4.4011e-02,  6.8400e-02, -6.5258e-02,  1.6776e-03,  4.7277e-02,
         -6.6036e-03, -7.3122e-03,  1.1962e-01,  2.2676e-03, -8.1218e-02,
         -8.7012e-02, -7.6099e-02,  1.2671e-01, -1.2231e-04,  2.8442e-02],
        [-3.8870e-02, -9.2241e-03, -2.0559e-01, -7.7822e-02,  8.9972e-02,
          7.7973e-02,  4.9886e-02, -7.0551e-02,  1.8816e-01, -1.8955e-02,
         -1.6270e-01,  4.4863e-03,  3.1283e-03, -4.3226e-02, -1.4738e-01,
         -8.7521e-04,  6.8936e-02,  

In [45]:
# @title get mean and relative weights
import torch as torch
mean_weights = weights.mean(dim=0)

abs_weights = torch.abs(mean_weights)
normalized_abs_weights = abs_weights / abs_weights.sum()
relative_weights = normalized_abs_weights * torch.sign(mean_weights)

print(relative_weights.shape)
print(relative_weights)


torch.Size([25])
tensor([-1.1105e-01,  1.5379e-02, -3.6414e-03, -1.5833e-02, -6.3142e-02,
         5.7702e-03,  1.0465e-02, -1.1539e-01,  5.6698e-02, -2.0032e-02,
        -2.0701e-02,  6.9248e-02,  4.7541e-02, -1.6587e-02,  4.1035e-02,
        -1.5231e-02, -1.5492e-02,  3.8261e-02, -1.1406e-02, -9.1052e-02,
         2.8389e-05,  6.9294e-02,  6.4406e-02,  1.0221e-03,  8.1294e-02])
tensor(3.5238e-05)


# Steering Experiments with Vectors

Given that we now have a list of coefficients to weight the steering vectors with, we want to find the actual steering vectors, weigh them, and create a vector representing our virtual feature.

In [42]:
from transformer_lens import HookedTransformer
model = HookedTransformer.from_pretrained(MODEL_NAME, tokenizer=tokenizer, device="cpu")

sae_out = sae.decode(sv_feature_acts)

hook_point = sae.cfg.hook_name

print("hook point")
print(hook_point)
print("------------------------------------")
print("model")
print(model)

Loaded pretrained model gpt2 into HookedTransformer
hook point
blocks.9.hook_resid_pre
------------------------------------
model
HookedTransformer(
  (embed): Embed()
  (hook_embed): HookPoint()
  (pos_embed): PosEmbed()
  (hook_pos_embed): HookPoint()
  (blocks): ModuleList(
    (0-11): 12 x TransformerBlock(
      (ln1): LayerNormPre(
        (hook_scale): HookPoint()
        (hook_normalized): HookPoint()
      )
      (ln2): LayerNormPre(
        (hook_scale): HookPoint()
        (hook_normalized): HookPoint()
      )
      (attn): Attention(
        (hook_k): HookPoint()
        (hook_q): HookPoint()
        (hook_v): HookPoint()
        (hook_z): HookPoint()
        (hook_attn_scores): HookPoint()
        (hook_pattern): HookPoint()
        (hook_result): HookPoint()
      )
      (mlp): MLP(
        (hook_pre): HookPoint()
        (hook_post): HookPoint()
      )
      (hook_attn_in): HookPoint()
      (hook_q_input): HookPoint()
      (hook_k_input): HookPoint()
      (hook_v_

In [91]:
# Code taken from SAELens Steering tutorial and modified
import einops

def steering_hook(resid_pre, hook):
    if resid_pre.shape[1] == 1:
        return

    position = sae_out.shape[1]
    if steering_on:
      # using our steering vector and applying the coefficient
      resid_pre[:, :position - 1, :] += steering_vector.to(device)


def hooked_generate(prompt_batch, fwd_hooks=[], seed=None, **kwargs):
    if seed is not None:
        torch.manual_seed(seed)

    with model.hooks(fwd_hooks=fwd_hooks):
        tokenized = model.to_tokens(prompt_batch)
        result = model.generate(
            stop_at_eos=False,  # avoids a bug on MPS
            input=tokenized,
            max_new_tokens=50,
            do_sample=True,
            **kwargs)
    return result

def run_generate(example_prompt):
  model.reset_hooks()
  editing_hooks = [(f"blocks.{layer}.hook_resid_post", steering_hook)]
  res = hooked_generate([example_prompt] * 3, editing_hooks, seed=None, **sampling_kwargs)

  # Print results, removing the ugly beginning of sequence token
  res_str = model.to_string(res[:, 1:])
  print(("\n\n" + "-" * 80 + "\n\n").join(res_str))

def create_average_steering_vector(feature_set, sae, multiplier):
  steering_vectors = torch.stack([sae.W_dec[feature_id] for feature_id in feature_set])
  coefficient_magic = (multiplier/len(steering_vectors))
  coefficients = torch.ones(len(steering_vectors))*coefficient_magic
  coefficients = coefficients.view(-1, 1)
  steering_vector = coefficients * steering_vectors
  steering_vector = torch.sum(steering_vector, dim=0)
  return steering_vector

def create_weighted_steering_vector(activation_dict, sae, weights, multiplier):
  steering_vectors = torch.stack([sae.W_dec[label] for label in feature_labels])
  steering_vector = einops.einsum(weights, steering_vectors, 'feat, feat d -> d')
  steering_vector = steering_vector * multiplier
  return steering_vector

In [75]:
tensor_list = list(activation_dict.values())
weights_tensor = torch.stack(tensor_list)

features_list = list(activation_dict.keys())

print(weights_tensor.shape)
print(features_list)

torch.Size([25, 30, 768])
[4612, 15879, 17803, 14219, 11536, 13087, 8994, 16174, 12210, 15423, 12745, 18638, 1871, 3665, 14940, 17244, 20445, 9185, 2787, 15870, 22003, 11764, 17012, 21369, 15227]


In [105]:
example_prompt = "On Mondays, I like to"
sampling_kwargs = dict(temperature=1.0, top_p=0.1, freq_penalty=1.0)

steering_vector = create_weighted_steering_vector(activation_dict, sae, relative_weights, 300)

steering_on = True
run_generate(example_prompt)

  0%|          | 0/50 [00:00<?, ?it/s]

On Mondays, I like to watch the latest episode of "The Simpsons" on Fox. It's a show that has been around for over 30 years and is one of the most popular shows on television. The show is a satire about how people are treated in America today, and

--------------------------------------------------------------------------------

On Mondays, I like to take a break from my usual work and spend time with my family. This week, I'm going to be doing something different.

I'm going to be taking a break from the work of writing for this blog because it's time for me

--------------------------------------------------------------------------------

On Mondays, I like to watch the NFL's most exciting games. The Seahawks are a team that has been playing great football for years and it's time to get back on track.

The Seahawks have won three of their last four games and they're looking to make history


In [106]:
sampling_kwargs = dict(temperature=1.0, top_p=0.1, freq_penalty=1.0)

steering_vector = create_average_steering_vector(activation_dict, sae, 300)

steering_on = True
run_generate(example_prompt)

  0%|          | 0/50 [00:00<?, ?it/s]

On Mondays, I like to talk about the things that make me happy.

I'm not talking about the things that make me feel good. I'm talking about the things that make me feel good in a way that makes you feel better.

I think it's

--------------------------------------------------------------------------------

On Mondays, I like to talk about the things that make me happy.

I'm not talking about the stuff that makes me feel good. I'm talking about the things that make me feel bad.

I'm talking about how much I love my job and how

--------------------------------------------------------------------------------

On Mondays, I like to think of myself as a guy who's always on the lookout for something new. I'm not a fan of the idea that there's anything wrong with me. But when it comes to my own life, I've found that there are things that make


In [107]:
steering_on = False
run_generate(example_prompt)

  0%|          | 0/50 [00:00<?, ?it/s]

On Mondays, I like to take a break from my usual work and spend time with my family. I'm not going to lie, it's been a great week for me.

I've been working on the project for about two weeks now and have had some great feedback

--------------------------------------------------------------------------------

On Mondays, I like to take a break from my usual routine of reading and writing. I'm not going to go into detail about what I do, but here are some things that I've learned over the years:

1. Don't be afraid to ask questions.

--------------------------------------------------------------------------------

On Mondays, I like to talk about the latest news and trends in the world of gaming. I'm not a big fan of "the old days" or "the new days", but I do think that there are some things that we can all learn from.

I


#Commented Out Code Blocks

Cells that we don't immediately need but may be useful later

In [ ]:
# @title Usage of clustering alg
# #@markdown Pick any layer, any feature from Joseph Bloom's GPT-2-small SAEs on the residual stream. Valid feature ids are between 0 and 24575.

# #@markdown `build_cluster` will return a list of features related to it, and a neuronpedia link to visualize of all of them.

# #@markdown If you use your own linkages for a different model, the features will still be related but the neuronpedia data won't be valid!

# #@markdown The `height` parameter controls how large the cluster is, by including more distant features.

# #@markdown If `height` is 6 or more, the URL might be too long to function.

# layer = "9" #@param [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
# feature_id = 2345 #@param {type: "integer"}
# height = 2 #@param {type: "slider", min:1, max:8}
# setting = 'average' #@param ['average', 'complete', 'weighted']
# indices = build_cluster(
#     layer=layer,
#     feature_id=feature_id,
#     height=height,
#     setting=setting,
# )

In [ ]:
# @title Specify dataset path
# dataset_path = ''

# with open(dataset_path, 'r') as f:
#     dataset = [json.loads(line) for line in f]

# # Create contrastive pairs
# contrastive_pairs = create_contrastive_pairs(dataset)

# # Print a sample pair
# print("Sample contrastive pair:")
# print("Positive prompt:", contrastive_pairs[0]["positive_prompt"])
# print("Positive completion:", contrastive_pairs[0]["positive_completion"])
# print("\nNegative prompt:", contrastive_pairs[0]["negative_prompt"])
# print("Negative completion:", contrastive_pairs[0]["negative_completion"])

In [ ]:
# @title helper methods related to the contrastive pair
# import random
# import json

# def create_contrastive_pairs(dataset, num_pairs=100):
#     contrastive_pairs = []

#     for data in dataset:
#         question = data["question"]
#         statement = data["statement"]
#         answer_matching = data["answer_matching_behavior"].strip()
#         answer_not_matching = data["answer_not_matching_behavior"].strip()

#         # Create positive example
#         positive_prompt = f"{question}\n{statement}\nA) {answer_matching}\nB) {answer_not_matching}\nAnswer:"
#         positive_completion = " A"

#         # Create negative example
#         negative_prompt = f"{question}\n{statement}\nA) {answer_matching}\nB) {answer_not_matching}\nAnswer:"
#         negative_completion = " B"

#         contrastive_pairs.append({
#             "positive_prompt": positive_prompt,
#             "positive_completion": positive_completion,
#             "negative_prompt": negative_prompt,
#             "negative_completion": negative_completion
#         })

#     # Ensure we have at least the requested number of pairs
#     while len(contrastive_pairs) < num_pairs:
#         contrastive_pairs.extend(contrastive_pairs)

#     # Randomly select the requested number of pairs
#     return random.sample(contrastive_pairs, num_pairs)


In [ ]:
# @title Linear Regression Function Definition

# import torch
# import einops
# from typing import Dict

# def fit_features_to_activation_delta(
#     feature_activations: Dict[int, torch.Tensor],
#     activation_delta: torch.Tensor
# ) -> torch.Tensor:
#     """
#     Fit a set of curated features to an activation delta using linear regression.

#     Args:
#     feature_activations (Dict[int, torch.Tensor]): Dictionary mapping feature indices to their activation tensors.
#     activation_delta (torch.Tensor): Target activation delta to fit.

#     Returns:
#     torch.Tensor: Weights that maximize similarity to the activation delta.
#     """
#     # Convert the dictionary to a list of tensors, preserving the order of indices
#     feature_indices = sorted(feature_activations.keys())
#     X = torch.stack([feature_activations[idx] for idx in feature_indices])

#     # Reshape X to (num_features, -1)
#     X = einops.rearrange(X, 'features ... -> features (...)')

#     # Reshape y (activation_delta) to (-1,)
#     y = einops.rearrange(activation_delta, '... -> (...)')

#     # Compute the weights using the normal equation
#     weights = torch.linalg.lstsq(X.T, y).solution

#     return weights

# # @title Getting regression results and weighing them
# for i in range(len(activation_deltas)):
#   X = fit_features_to_activation_delta(activation_dict, activation_deltas[i, filtered_indices, :])
#   print(X)